README.md
# recipe-app-api

[![ Build Status](https://travis-ci.com/kaychess/recipe-app-api2.svg?branch=master)](https://travis-ci.com/kaychess/recipe-app-api2)


CHANGELOG.md
# Semantic Versioning Changelog

# [3.1.0](https://github.com/kaychess/recipe-app-api2/compare/v3.0.0...v3.1.0) (2020-06-29)


### Features

* **models:** added custom user model ([314be46](https://github.com/kaychess/recipe-app-api2/commit/314be46c8fc6016b7311eed5609d3ff4bbaa782a))

# [3.0.0](https://github.com/kaychess/recipe-app-api2/compare/v2.0.0...v3.0.0) (2020-06-29)


### Build System

* **git actions:** change git commit message linting from travis ci to github actions ([123582d](https://github.com/kaychess/recipe-app-api2/commit/123582d4bcd0b90dfde9c77f095b6541b2ec6dad))


### BREAKING CHANGES

* **git actions:** changed the git commit message linting from travis ci build systems to github
actions

# [2.0.0](https://github.com/kaychess/recipe-app-api2/compare/v1.0.0...v2.0.0) (2020-06-29)


### Build System

* change commit message lint and semantic release to git actions ([8617a20](https://github.com/kaychess/recipe-app-api2/commit/8617a20f9fd392cfa086eaa3fc411071eab2b53d))


### BREAKING CHANGES

* git actions would take care of commit message linting rather to travis CI

# 1.0.0 (2020-06-29)


### Bug Fixes

* init ([599f775](https://github.com/kaychess/recipe-app-api2/commit/599f7759c32d11713ab9f6227d7d5c06bf2269de))
* initial commit ([77483c7](https://github.com/kaychess/recipe-app-api2/commit/77483c7fc4c30ae2baf25d5e2486df1b881e28b5))
* testing semantic versioning ([3cd8736](https://github.com/kaychess/recipe-app-api2/commit/3cd873661906d47e2be19c2cdf09ec1ac38924ae))
* testing semantic versioning ([8ed51fb](https://github.com/kaychess/recipe-app-api2/commit/8ed51fb1f0ec7af7f44ffd92b9e8b34003613e94))

# [3.0.0](https://github.com/kaychess/recipe-app-api/compare/v2.0.1...v3.0.0) (2020-06-29)


### Bug Fixes

* **test:** want to test what happens when bot doesnt build semantic relase ([2265ae7](https://github.com/kaychess/recipe-app-api/commit/2265ae76359a7d697b86369b53a54f868511d5cf))


### BREAKING CHANGES

* **test:** just a test dont worry#1

# [3.0.0](https://github.com/kaychess/recipe-app-api/compare/v2.0.1...v3.0.0) (2020-06-29)


### Bug Fixes

* **test:** want to test what happens when bot doesnt build semantic relase ([2265ae7](https://github.com/kaychess/recipe-app-api/commit/2265ae76359a7d697b86369b53a54f868511d5cf))


### BREAKING CHANGES

* **test:** just a test dont worry#1

# [3.0.0](https://github.com/kaychess/recipe-app-api/compare/v2.0.1...v3.0.0) (2020-06-29)


### Bug Fixes

* **test:** want to test what happens when bot doesnt build semantic relase ([2265ae7](https://github.com/kaychess/recipe-app-api/commit/2265ae76359a7d697b86369b53a54f868511d5cf))


### BREAKING CHANGES

* **test:** just a test dont worry#1

# [3.0.0](https://github.com/kaychess/recipe-app-api/compare/v2.0.1...v3.0.0) (2020-06-29)


### Bug Fixes

* **test:** want to test what happens when bot doesnt build semantic relase ([2265ae7](https://github.com/kaychess/recipe-app-api/commit/2265ae76359a7d697b86369b53a54f868511d5cf))


### BREAKING CHANGES

* **test:** just a test dont worry#1

# [3.0.0](https://github.com/kaychess/recipe-app-api/compare/v2.0.1...v3.0.0) (2020-06-29)


### Bug Fixes

* **test:** want to test what happens when bot doesnt build semantic relase ([2265ae7](https://github.com/kaychess/recipe-app-api/commit/2265ae76359a7d697b86369b53a54f868511d5cf))


### BREAKING CHANGES

* **test:** just a test dont worry#1

# [3.0.0](https://github.com/kaychess/recipe-app-api/compare/v2.0.1...v3.0.0) (2020-06-29)


### Bug Fixes

* **test:** want to test what happens when bot doesnt build semantic relase ([2265ae7](https://github.com/kaychess/recipe-app-api/commit/2265ae76359a7d697b86369b53a54f868511d5cf))


### BREAKING CHANGES

* **test:** just a test dont worry#1

# [3.0.0](https://github.com/kaychess/recipe-app-api/compare/v2.0.1...v3.0.0) (2020-06-29)


### Bug Fixes

* **test:** want to test what happens when bot doesnt build semantic relase ([2265ae7](https://github.com/kaychess/recipe-app-api/commit/2265ae76359a7d697b86369b53a54f868511d5cf))


### BREAKING CHANGES

* **test:** just a test dont worry#1

# [3.0.0](https://github.com/kaychess/recipe-app-api/compare/v2.0.1...v3.0.0) (2020-06-29)


### Bug Fixes

* **test:** want to test what happens when bot doesnt build semantic relase ([2265ae7](https://github.com/kaychess/recipe-app-api/commit/2265ae76359a7d697b86369b53a54f868511d5cf))


### BREAKING CHANGES

* **test:** just a test dont worry#1

# [3.0.0](https://github.com/kaychess/recipe-app-api/compare/v2.0.1...v3.0.0) (2020-06-29)


### Bug Fixes

* **test:** want to test what happens when bot doesnt build semantic relase ([2265ae7](https://github.com/kaychess/recipe-app-api/commit/2265ae76359a7d697b86369b53a54f868511d5cf))


### BREAKING CHANGES

* **test:** just a test dont worry#1

## [2.0.1](https://github.com/kaychess/recipe-app-api/compare/v2.0.0...v2.0.1) (2020-06-24)


### Bug Fixes

* **ci cd update:** modify @commitlint/config-angular to config-conventional ([5bb5074](https://github.com/kaychess/recipe-app-api/commit/5bb5074bcf9c2a306e38169123835f3a9f8b0672))

# [2.0.0](https://github.com/kaychess/recipe-app-api/compare/v1.6.2...v2.0.0) (2020-06-24)


### Features

* **ci cd updated:** add git commit lint and schematic release ([7d76597](https://github.com/kaychess/recipe-app-api/commit/7d765972e30b8418570a2e31c55ca954f7ce98f1)), closes [#1](https://github.com/kaychess/recipe-app-api/issues/1)


### BREAKING CHANGES

* **ci cd updated:** enforces to use angular conventional conventions for git commit messages and generates semantic relases automatically

## [1.6.2](https://github.com/kaychess/recipe-app-api/compare/v1.6.1...v1.6.2) (2020-06-24)


### Bug Fixes

* final check if semantic release is working ([12dfe89](https://github.com/kaychess/recipe-app-api/commit/12dfe89d46ed4bc3e2374ea9f7e12581d1ae7ba5))

## [1.6.1](https://github.com/kaychess/recipe-app-api/compare/v1.6.0...v1.6.1) (2020-06-24)


### Bug Fixes


app/core/tests/test_models.py
from django.test import TestCase
# it's recommended to use get_user_model rather to user model directly
# for flexibility of changing user model later in settings.
from django.contrib.auth import get_user_model


class TestUserModel (TestCase):
    ''' Test if, Can we create a User '''

    def test_create_user_with_email_successfully(self):
        # need NOT be a valid email or password.
        email = 'test@menuroma.com'
        password = 'ABC@123'

        # Creates a user using the configured user model.
        # Note : object with data is created, but not saved yet.
        user = get_user_model().objects.create_user(
            email=email,
            password=password
        )

        self.assertEqual(user.email, email)
        self.assertTrue(user.check_password(password))

    def test_new_user_normalized(self):
        ''' Test the email for a new user is normalized '''
        email = 'test@MENUROMA.COM'
        password = 'ABC@123'

        user = get_user_model().objects.create_user(
            email=email,
            password=password
        )

        self.assertEqual(user.email, email.lower())

    def test_new_user_invalid(self):
        ''' Test creating user with no email raises error '''
        # Anything that runs in 'with' block has to raise a value error
        with self.assertRaises(ValueError):
            get_user_model().objects.create_user(
                email=None,
                password='ABC@123')

    def test_create_new_super_user(self):
        ''' Test creating new super user '''
        # need NOT be a valid email or password.
        email = 'superUser@menuroma.com'
        password = 'ABC@123'
        user = get_user_model().objects.create_superuser(
            email=email,
            password=password
        )

        self.assertTrue(user.is_superuser)
        self.assertTrue(user.is_staff)


app/core/models.py
from django.db import models
from django.contrib.auth.models import BaseUserManager, AbstractBaseUser, \
    PermissionsMixin


class UserManager(BaseUserManager):
    def create_user(self, email, password=None, **extra_fields):
        ''' Creates and saves new user '''
        if not email:
            raise ValueError('Users must have an email address')
        user = self.model(email=self.normalize_email(email), **extra_fields)
        user.set_password(password)
        user.save(using=self._db)

        return user

    def create_superuser(self, email, password):
        '''Create and saves new super user '''
        user = self.create_user(email, password)
        user.is_staff = True
        user.is_superuser = True
        user.save(using=self._db)

        return user


class User(AbstractBaseUser, PermissionsMixin):
    '''Custom user model that supports using email instead of username '''
    email = models.EmailField(max_length=255, unique=True)
    name = models.CharField(max_length=255)
    is_active = models.BooleanField(default=True)
    is_staff = models.BooleanField(default=False)

    objects = UserManager()

    USERNAME_FIELD = 'email'


app/app/settings.py
"""
Django settings for app project.

Generated by 'django-admin startproject' using Django 3.0.7.

For more information on this file, see
https://docs.djangoproject.com/en/3.0/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/3.0/ref/settings/
"""

import os

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/3.0/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = 'l19t!r24rf1a5uojsqw5w0^@xow%_3i!(k=nxt@p00521u5fel'

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True

ALLOWED_HOSTS = []


# Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'core',
]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

ROOT_URLCONF = 'app.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'app.wsgi.application'


# Database
# https://docs.djangoproject.com/en/3.0/ref/settings/#databases

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
    }
}


# Password validation
# https://docs.djangoproject.com/en/3.0/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]


# Internationalization
# https://docs.djangoproject.com/en/3.0/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/3.0/howto/static-files/

STATIC_URL = '/static/'


AUTH_USER_MODEL = 'core.User'


app/core/migrations/0001_initial.py
# Generated by Django 3.0.7 on 2020-06-29 23:08

from django.db import migrations, models


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        ('auth', '0011_update_proxy_permissions'),
    ]

    operations = [
        migrations.CreateModel(
            name='User',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('password', models.CharField(max_length=128, verbose_name='password')),
                ('last_login', models.DateTimeField(blank=True, null=True, verbose_name='last login')),
                ('is_superuser', models.BooleanField(default=False, help_text='Designates that this user has all permissions without explicitly assigning them.', verbose_name='superuser status')),
                ('email', models.EmailField(max_length=255, unique=True)),
                ('name', models.CharField(max_length=255)),
                ('is_active', models.BooleanField(default=True)),
                ('is_staff', models.BooleanField(default=False)),
                ('groups', models.ManyToManyField(blank=True, help_text='The groups this user belongs to. A user will get all permissions granted to each of their groups.', related_name='user_set', related_query_name='user', to='auth.Group', verbose_name='groups')),
                ('user_permissions', models.ManyToManyField(blank=True, help_text='Specific permissions for this user.', related_name='user_set', related_query_name='user', to='auth.Permission', verbose_name='user permissions')),
            ],
            options={
                'abstract': False,
            },
        ),
    ]
